### PR TITLE
replace boost with std variant

### PR DESCRIPTION
**Pull request by timosachsenberg**
# Description

Replaces boost::variant with std::variant (which has slightly different semantics regarding comparison operators)

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- /rebase will try to rebase the PR on the current develop branch.
- /reformat (experimental) applies the clang-format style changes as additional commit
- setting the label NoJenkins will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)

### CODE REVIEW

**Comment by timosachsenberg** [comment]
btw now most output in the log is generated from:

checking (MorphologicalFilter()) ... 
: passed

checking (virtual ~MorphologicalFilter()) ... : passed
Warning: no subtests performed in '(virtual ~MorphologicalFilter())' (line 252)!

checking (template < typename InputIterator, typename OutputIterator > void filterRange( InputIterator input_begin, InputIterator input_end, OutputIterator output_begin)) ... 
    line 288: status:  data_size: 1
    line 289: status:  struc_elem_length: 3
    line 291: status:  erosion
    line 312: status:  0
    line 318: status:  dilation
    line 339: status:  0
    line 288: status:  data_size: 2
    line 289: status:  struc_elem_length: 3
    line 291: status:  erosion
    line 312: status:  0
    line 312: status:  1
    line 318: status:  dilation
    line 339: status:  0
    line 339: status:  1
    line 288: status:  data_size: 2
    line 289: status:  struc_elem_length: 5
    line 291: status:  erosion
    line 312: status:  0
    line 312: status:  1
...
```

**Comment by jpfeuffer** [comment]
I guess in std, the operators are defined outside of the class. Therefore one way to use it is to also define the overwriting ones outside of the class:

```
typedef std::variant<int, double> RefVariant;

struct IdentifiedMolecule: public RefVariant
{
  IdentifiedMolecule() = default;
  IdentifiedMolecule(int ref): RefVariant(ref) {};
  IdentifiedMolecule(double ref): RefVariant(ref) {};
  IdentifiedMolecule(const IdentifiedMolecule&) = default;

};
bool operator==(const IdentifiedMolecule& one, const IdentifiedMolecule& other)
{
    return operator==(static_cast<RefVariant>(one), static_cast<RefVariant>(other));
}
bool operator!=(const IdentifiedMolecule& one, const IdentifiedMolecule& other)
{
    return !operator==(one, other);
}
bool operator<(const IdentifiedMolecule& one, const IdentifiedMolecule& other)
{
    return operator<(static_cast<RefVariant>(one), static_cast<RefVariant>(other));
}
```

File: src/openms/source/METADATA/ID/IdentifiedMolecule.cpp [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+      return getIdentifiedCompoundRef()->identifier; // or use "name"?
+    case MoleculeType::RNA:
+      return getIdentifiedOligoRef()->sequence.toString();
+    default:
+      throw Exception::NotImplemented(__FILE__, __LINE__,
+                                      OPENMS_PRETTY_FUNCTION);
+  }
+}
+
+} // namespace
```
Comment (timosachsenberg) [inline src/openms/source/METADATA/ID/IdentifiedMolecule.cpp:147]:
```suggestion
} // namespace

```

**Review by hendrikweisser — APPROVED** [review]
Looks good! Thanks for making the changes. (I don't think any of the uses of `std::get_if` could be simplified by using `std::get`, either.)

My only question is: Were changes in `MorphologicalFilter_test` intended to be part of this PR?

**Review by jpfeuffer — APPROVED** [review]

**Comment by timosachsenberg** [comment]
These we’re part of another PR. Should go away if I merge the other first

**Pull request merged**